Improve src_test documentation

mgorny opened the pull request:
As noted in gentoo/gentoo#4544, we have some old stuff there.

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,13 +59,13 @@ src_test() {
-    # Try to run the non-gui tests only
-    make test-nongui \
+    # Try to run the non-gui tests only, non-parallel
+    emake -j1 test-nongui \
```
If you switch to `emake`, do you still need `|| die` there?

mgorny replied:
Depends on EAPI, obviously.

gktrk replied:
Let's assume EAPI5+

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,13 +59,13 @@ src_test() {
 
-    # Try to run the non-gui tests only
-    make test-nongui \
+    # Try to run the non-gui tests only, non-parallel
```
Should we rather say `for non-parallelism`? I'm guessing this is related to the hunk about parallelism you deleted below.

mgorny replied:
Sounds like a weird word. It's just not run in parallel.

kentfredric replied:
> Try to run the non-gui tests only, single threaded

?

gktrk replied:
That works too

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -82,7 +82,6 @@ portage environment. Reasons for such a failure could include:
 </p>
 
 <ul>
-  <li>Needing to use X.</li>
```
Is this related to the comment `patched out the tests that don't work without X.` in PR 4544? Instead of removing this sentence, should we expand it with this?

```
Reasons for such a failure could include:
- Needing to use X (Unless the test system can be configured/patched to exclude those certain tests).
```

mgorny replied:
Tests needing to use X should be run via virtualx.eclass.

gktrk replied:
Why not put that information in there instead?

mgorny replied:
Feel free to do that. Last I checked, I wasn't on devmanual team or responsible for fixing horribly bad documentation.

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -96,12 +95,6 @@ using <c>sed</c> or skipping a particular <c>make</c> target is
-parallelise tests unless the <c>Makefile</c> was specifically designed
-for this can cause all sorts of strange problems.
-</note>
-
```
I like the direction of saying `use emake -j1` instead of `do not use emake`. I wonder if we should update this sentence and make it compliant with the example above or simple remove it. Not sure personally.

mgorny replied:
I don't think there's a point in making a special verbose notes about past state not applying anymore.

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -114,26 +107,15 @@ suite is extremely helpful for arch maintainers.
-indicates a genuine fault.
+to them about getting it fixed. A broken test suite requires developers
+to investigate each test failure in order to determine whether it is
+genuine or indicates a broken test.
```
I personally like the old version better. Likewise a broken test suite, a non-broken test suite also requires developers to investigate each test failure since there's no way `to tell whether a test failure indicates a genuine fault.`

mgorny replied:
If test suite is not broken, then a failure indicates a genuine problem, obviously.

The old version stated falsely that 'we are unable to tell [...]' — we *are* able to tell that, it just requires additional effort.

mgorny commented:
Updated the non-parallel comment and removed die from emake.

mgorny commented:
@gentoo/devmanual, please review now. Updated the obsolete snippet from Portage to a modern version, and added detailed section on network/service access and on dealing with X11 (which is basically a specific example of the generic local service problem).

kensington commented:
These new changes are really, really nice. Good work!

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,14 +49,161 @@ src_test() {
+  </li>
+
+  <li>
+    using production services for running tests is frequently
```
s/frequently/extremely/

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,14 +49,161 @@ src_test() {
+</ul>
+
+<p>
+Fixing tests that require access to the local services is usually done
```
s/the local services/local services/

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,14 +49,161 @@ src_test() {
+
+<p>
+For all networked services exposed during the test phase (either by
+the ebuild or the tests itself), UNIX sockets are strongly preferred
```
`tests themselves` seems more correct to me

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,14 +49,161 @@ src_test() {
+the ebuild or the tests itself), UNIX sockets are strongly preferred
+over IP sockets as they provide better means for unique naming
+and access control mechanisms. The IP sockets can be subject to port
+collisions with other local services and they can be connected to
```
s/connected to/accessed/

gktrk commented on ebuild-writing/functions/src_test/text.xml:
s/the developers/developers/

gktrk commented on ebuild-writing/functions/src_test/text.xml:
Either `ensure that this variable is unset` or `simply unset this variable`

gktrk commented on ebuild-writing/functions/src_test/text.xml:
s/devices/device/

gktrk commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,14 +49,161 @@ src_test() {
+
+  <li>
+    accessing remote sites always poses a privacy issue, and possibly
+    even a threat to security (e.g. through inadvertently exposing
```
s/even//

mgorny commented on ebuild-writing/functions/src_test/text.xml:
```diff
@@ -59,14 +49,161 @@ src_test() {
+Fixing tests that require Internet access usually requires cooperation
+with upstream, and porting the tests to use one of the test techniques
+such as mocking or using replay data. For this reason, developers
+more frequently report the issue upstream and skip tests that require
```
This changes the meaning of the sentence completely.

mgorny commented:
Updated.

gktrk commented:
http://dev.gentoo.org/~gokturk/devmanual/pr65/ebuild-writing/functions/src_test/index.html to check how it renders.

gktrk commented:
Merged in https://github.com/gentoo/devmanual.gentoo.org/compare/9456029...1118c72. Thanks!

I've mangled the summary line prefix before merging.
